Transports can't create directories when copying files

### Describe the bug

Both the `local` and `ssh` transport except when folders need to be created for the full path of a copied file. This manifest itself when paths like `dirname/filename.txt` are added to the copy list of calcjobs and the `dirname` folder does not already exist.

### Steps to reproduce

1. Set up a remote computer with `ssh` or `local` transport and configure it. In my case, I have `daint` for `ssh` and `localhost` for `local` transport.
2. Set up a code on those computers by running the following command: `verdi code setup -L bash -P templatereplacer --on-computer --computer=daint --remote-abs-path=/bin/bash --non-interactive` (the only flag you need to customize is the `--computer=localhost`)
3. Run (`verdi run`) the following script:

```
from aiida.orm import Dict, SinglefileData, load_code
from aiida.engine import run_get_node

import tempfile
with tempfile.NamedTemporaryFile(mode='w+') as handle:
    filepath = handle.name
    handle.write('content of file')
    handle.flush()
    sfile_node = SinglefileData(file=filepath).store()

param_dict = Dict(dict={'x': 3, 'y': 4}).store()
templ_dict = Dict(dict={
    'input_file_name': 'script.sh',
    'input_file_template': "echo $(({x} * {y}))",
    'cmdline_params': ['script.sh'],
    'output_file_name': 'stdout.txt',
    #'files_to_copy': [('inpfile','file.txt')],
    #'files_to_copy': [('inpfile','dir1/file.txt')],
    }).store()

code = load_code('bash@daint')
#code = load_code('bash@localhost')

builder = code.get_builder()
builder.parameters = param_dict
builder.template = templ_dict
builder.files = {'inpfile': sfile_node}
builder.metadata.options = {
    'resources': {'num_machines': 1, 'tot_num_mpiprocs': 1},
    'custom_scheduler_commands': '#SBATCH --constraint=mc\n#SBATCH --cpus-per-task=36',
    'max_wallclock_seconds': 60,
    }

print('param dict:  {}'.format(param_dict.uuid))
print('templ dict:  {}'.format(templ_dict.uuid))
print('sfile node:  {}'.format(sfile_node.uuid))

results, calcjobnode = run_get_node(builder)
print('calculation: {}'.format(calcjobnode.uuid))
```

You can change the two commented sections. The first one lets you check that setting a path that doesn't need to create folders will work. The second one is for testing the `ssh` or `local` transports and see they behave the same.

### Expected behavior

I would expect that all necessary directories would be automatically created if they don't exist. This is because there are some programs that will require a certain file structure, and there is no independent way of just creating folders.

Instead I get this error message for `ssh` transport:

```
09/04/2020 11:17:51 AM <2377> aiida.engine.transports: [ERROR] Exception whilst using transport:
Traceback (most recent call last):
  File "/home/framirez/Workgit/aiida-core/aiida/engine/transports.py", line 103, in request_transport
    yield transport_request.future
  File "/home/framirez/Workgit/aiida-core/aiida/engine/processes/calcjobs/tasks.py", line 83, in do_upload
    execmanager.upload_calculation(node, transport, calc_info, folder)
  File "/home/framirez/Workgit/aiida-core/aiida/engine/daemon/execmanager.py", line 205, in upload_calculation
    transport.put(handle.name, target)
  File "/home/framirez/Workgit/aiida-core/aiida/transports/plugins/ssh.py", line 765, in put
    self.putfile(localpath, remotepath, callback, dereference, overwrite)
  File "/home/framirez/Workgit/aiida-core/aiida/transports/plugins/ssh.py", line 792, in putfile
    return self.sftp.put(localpath, remotepath, callback=callback)
  File "/home/framirez/.virtualenvs/aiida-core/lib/python3.6/site-packages/paramiko/sftp_client.py", line 759, in put
    return self.putfo(fl, remotepath, file_size, callback, confirm)
  File "/home/framirez/.virtualenvs/aiida-core/lib/python3.6/site-packages/paramiko/sftp_client.py", line 714, in putfo
    with self.file(remotepath, "wb") as fr:
  File "/home/framirez/.virtualenvs/aiida-core/lib/python3.6/site-packages/paramiko/sftp_client.py", line 372, in open
    t, msg = self._request(CMD_OPEN, filename, imode, attrblock)
  File "/home/framirez/.virtualenvs/aiida-core/lib/python3.6/site-packages/paramiko/sftp_client.py", line 813, in _request
    return self._read_response(num)
  File "/home/framirez/.virtualenvs/aiida-core/lib/python3.6/site-packages/paramiko/sftp_client.py", line 865, in _read_response
    self._convert_status(msg)
  File "/home/framirez/.virtualenvs/aiida-core/lib/python3.6/site-packages/paramiko/sftp_client.py", line 894, in _convert_status
    raise IOError(errno.ENOENT, text)
FileNotFoundError: [Errno 2] No such file
```

And this one for `local` transport:

```
09/04/2020 11:07:39 AM <1447> aiida.engine.transports: [ERROR] Exception whilst using transport:
Traceback (most recent call last):
  File "/home/framirez/Workgit/aiida-core/aiida/engine/transports.py", line 103, in request_transport
    yield transport_request.future
  File "/home/framirez/Workgit/aiida-core/aiida/engine/processes/calcjobs/tasks.py", line 83, in do_upload
    execmanager.upload_calculation(node, transport, calc_info, folder)
  File "/home/framirez/Workgit/aiida-core/aiida/engine/daemon/execmanager.py", line 205, in upload_calculation
    transport.put(handle.name, target)
  File "/home/framirez/Workgit/aiida-core/aiida/transports/plugins/local.py", line 296, in put
    self.putfile(localpath, full_destination, overwrite)
  File "/home/framirez/Workgit/aiida-core/aiida/transports/plugins/local.py", line 333, in putfile
    shutil.copyfile(localpath, the_destination)
  File "/home/framirez/.virtualenvs/aiida-core/lib/python3.6/shutil.py", line 121, in copyfile
    with open(dst, 'wb') as fdst:
FileNotFoundError: [Errno 2] No such file or directory: '/scratch/framirez/aiida/49/17/2007-d8be-48f5-8c9b-80c1d7d222ff/dir1/file.txt'
```

### Your environment

- OS: Ubuntu 18.04.5 LTS
- Python 3.6.9
- aiida-core develop branch, commit 7c31bc2d54df8ee22db2fd20f6078a44ba871a89.

--- a/aiida/engine/daemon/execmanager.py
+++ b/aiida/engine/daemon/execmanager.py
@@ -206,27 +206,6 @@ def upload_calculation(node, transport, calc_info, folder, inputs=None, dry_run=
             logger.debug('[submission of calculation {}] copying file/folder {}...'.format(node.pk, filename))
             transport.put(folder.get_abs_path(filename), filename)
 
-    if dry_run:
-        if remote_copy_list:
-            with open(os.path.join(workdir, '_aiida_remote_copy_list.txt'), 'w') as handle:
-                for remote_computer_uuid, remote_abs_path, dest_rel_path in remote_copy_list:
-                    handle.write(
-                        'would have copied {} to {} in working directory on remote {}'.format(
-                            remote_abs_path, dest_rel_path, computer.label
-                        )
-                    )
-
-        if remote_symlink_list:
-            with open(os.path.join(workdir, '_aiida_remote_symlink_list.txt'), 'w') as handle:
-                for remote_computer_uuid, remote_abs_path, dest_rel_path in remote_symlink_list:
-                    handle.write(
-                        'would have created symlinks from {} to {} in working directory on remote {}'.format(
-                            remote_abs_path, dest_rel_path, computer.label
-                        )
-                    )
-
-    else:
-
         for (remote_computer_uuid, remote_abs_path, dest_rel_path) in remote_copy_list:
             if remote_computer_uuid == computer.uuid:
                 logger.debug(
@@ -268,6 +247,25 @@ def upload_calculation(node, transport, calc_info, folder, inputs=None, dry_run=
                     'It is not possible to create a symlink between two different machines for '
                     'calculation {}'.format(node.pk)
                 )
+    else:
+
+        if remote_copy_list:
+            with open(os.path.join(workdir, '_aiida_remote_copy_list.txt'), 'w') as handle:
+                for remote_computer_uuid, remote_abs_path, dest_rel_path in remote_copy_list:
+                    handle.write(
+                        'would have copied {} to {} in working directory on remote {}'.format(
+                            remote_abs_path, dest_rel_path, computer.label
+                        )
+                    )
+
+        if remote_symlink_list:
+            with open(os.path.join(workdir, '_aiida_remote_symlink_list.txt'), 'w') as handle:
+                for remote_computer_uuid, remote_abs_path, dest_rel_path in remote_symlink_list:
+                    handle.write(
+                        'would have created symlinks from {} to {} in working directory on remote {}'.format(
+                            remote_abs_path, dest_rel_path, computer.label
+                        )
+                    )
 
     # Loop recursively over content of the sandbox folder copying all that are not in `provenance_exclude_list`. Note
     # that directories are not created explicitly. The `node.put_object_from_filelike` call will create intermediate

--- a/aiida/engine/daemon/execmanager.py
+++ b/aiida/engine/daemon/execmanager.py
@@ -14,6 +14,7 @@ the routines make reference to the suitable plugins for all
 plugin-specific operations.
 """
 import os
+import shutil
 
 from aiida.common import AIIDA_LOGGER, exceptions
 from aiida.common.folders import SandboxFolder
@@ -153,17 +154,12 @@ def upload_calculation(node, transport, calc_info, folder, inputs=None, dry_run=
                     transport.put(handle.name, filename)
             transport.chmod(code.get_local_executable(), 0o755)  # rwxr-xr-x
 
-    # In a dry_run, the working directory is the raw input folder, which will already contain these resources
-    if not dry_run:
-        for filename in folder.get_content_list():
-            logger.debug('[submission of calculation {}] copying file/folder {}...'.format(node.pk, filename))
-            transport.put(folder.get_abs_path(filename), filename)
-
     # local_copy_list is a list of tuples, each with (uuid, dest_rel_path)
     # NOTE: validation of these lists are done inside calculation.presubmit()
     local_copy_list = calc_info.local_copy_list or []
     remote_copy_list = calc_info.remote_copy_list or []
     remote_symlink_list = calc_info.remote_symlink_list or []
+    provenance_exclude_list = calc_info.provenance_exclude_list or []
 
     for uuid, filename, target in local_copy_list:
         logger.debug('[submission of calculation {}] copying local file/folder to {}'.format(node.uuid, target))
@@ -196,13 +192,19 @@ def upload_calculation(node, transport, calc_info, folder, inputs=None, dry_run=
         if data_node is None:
             logger.warning('failed to load Node<{}> specified in the `local_copy_list`'.format(uuid))
         else:
-            # Note, once #2579 is implemented, use the `node.open` method instead of the named temporary file in
-            # combination with the new `Transport.put_object_from_filelike`
-            # Since the content of the node could potentially be binary, we read the raw bytes and pass them on
-            with NamedTemporaryFile(mode='wb+') as handle:
-                handle.write(data_node.get_object_content(filename, mode='rb'))
-                handle.flush()
-                transport.put(handle.name, target)
+            dirname = os.path.dirname(target)
+            if dirname:
+                os.makedirs(os.path.join(folder.abspath, dirname), exist_ok=True)
+            with folder.open(target, 'wb') as handle:
+                with data_node.open(filename, 'rb') as source:
+                    shutil.copyfileobj(source, handle)
+            provenance_exclude_list.append(target)
+
+    # In a dry_run, the working directory is the raw input folder, which will already contain these resources
+    if not dry_run:
+        for filename in folder.get_content_list():
+            logger.debug('[submission of calculation {}] copying file/folder {}...'.format(node.pk, filename))
+            transport.put(folder.get_abs_path(filename), filename)
 
     if dry_run:
         if remote_copy_list:
@@ -266,8 +268,6 @@ def upload_calculation(node, transport, calc_info, folder, inputs=None, dry_run=
                     'It is not possible to create a symlink between two different machines for '
                     'calculation {}'.format(node.pk)
                 )
-
-    provenance_exclude_list = calc_info.provenance_exclude_list or []
 
     # Loop recursively over content of the sandbox folder copying all that are not in `provenance_exclude_list`. Note
     # that directories are not created explicitly. The `node.put_object_from_filelike` call will create intermediate

--- a/tests/engine/test_launch.py
+++ b/tests/engine/test_launch.py
@@ -31,7 +31,9 @@ class FileCalcJob(CalcJob):
     def prepare_for_submission(self, folder):
         from aiida.common.datastructures import CalcInfo, CodeInfo
 
-        local_copy_list = [(self.inputs.single_file.uuid, self.inputs.single_file.filename, 'single_file')]
+        # Use nested path for the target filename, where the directory does not exist, to check that the engine will
+        # create intermediate directories as needed. Regression test for #4350
+        local_copy_list = [(self.inputs.single_file.uuid, self.inputs.single_file.filename, 'path/single_file')]
 
         for name, node in self.inputs.files.items():
             local_copy_list.append((node.uuid, node.filename, name))
@@ -286,5 +288,5 @@ class TestLaunchersDryRun(AiidaTestCase):
 
         _, node = launch.run_get_node(FileCalcJob, **inputs)
         self.assertIn('folder', node.dry_run_info)
-        for filename in ['single_file', 'file_one', 'file_two']:
+        for filename in ['path', 'file_one', 'file_two']:
             self.assertIn(filename, os.listdir(node.dry_run_info['folder']))

